R/exploratoryfactoranalysis.R

```diff
@@ -413,6 +491,70 @@ ExploratoryFactorAnalysis <- function(jaspResults, dataset, options, ...) {
 
 }
 
+# Modification here:
+# Creates a table that outputs the details of the parallel analysis.
+# Particularly useful for reporting the rationale of how many factors were
+# retained - especially within psychometric papers.
+# We are also going to be adding an asterisk aside any factor whose real data eigenvalue is above
+# the 95th percentile simulated data eigenvalue
+.efaPATable <- function(modelContainer, dataset, options, ready) {
+  if (!options[["incl_PAtable"]] || !is.null(modelContainer[["patab"]])) return()
+
+  if (options[["basedOn"]] == "mixed") {
+    polytetracor <- psych::mixedCor(dataset)
+    pa <- try(psych::fa.parallel(polytetracor$rho,
+                                 plot = FALSE,
+                                 fa = options$parallelMethod,
+                                 n.obs = nrow(dataset)))
+  }
+  else {
+    pa <- try(psych::fa.parallel(dataset, plot = FALSE, fa = options$parallelMethod))
+  }
+
+  patab <- createJaspTable(gettext("Parallel Analysis"))
+  patab$dependOn("incl_PAtable")
+  patab$addColumnInfo(name = "col", title = "", type = "string")
+  patab$addColumnInfo(name = "val1", title = gettext("Real Data Eigenvalue"), type = "number", format = "dp:3")
+  patab$addColumnInfo(name = "val2", title = gettext("Mean of Resampled Eigenvalues"), type = "number", format = "dp:3")
+  patab$position <- 5
+  modelContainer[["patab"]] <- patab
+
+  if (!ready || modelContainer$getError()) return()
+
+  if (options$parallelMethod == "pc") {
+    RealDataEigen <- pa$pc.values
+    ResampledEigen <- pa$pc.sim
+  } else { # parallelmethod is fa
+    RealDataEigen <- pa$fa.values
+    ResampledEigen <- pa$fa.sim
+  }
+
+  efaResult <- modelContainer[["model"]][["object"]]
+  PAs <- zapsmall(as.matrix(data.frame(RealDataEigen,ResampledEigen),fix.empty.names = F))
+  forasterisk <- data.frame(applyforasterisk=RealDataEigen-ResampledEigen)
+  dims <- nrow(PAs)
+
+  firstcol <- data.frame()
+  for (i in 1:dims) {
+    if (forasterisk$applyforasterisk[i] > 0) {
+      firstcol <- rbind(firstcol,paste("Factor", " ", i,"*", sep=""))
+    }
+    else {
+      firstcol <- rbind(firstcol,paste("Factor", i))
+    }
+  }
+  patab[["col"]] <- c(firstcol$X.Factor.1..)
+  patab[["val1"]] <- c(RealDataEigen)
+  patab[["val2"]] <- c(ResampledEigen)
+
+
+  if (options$parallelMethod == "pc") {
+    patab$addFootnote(message = gettext("* = Advised factor to be retained.\nResults from PC-based parallel analysis"))
+  } else { # parallelmethod is fa
+    patab$addFootnote(message = gettext("* = Advised factor to be retained.\nResults from FA-based parallel analysis"))
+  }
```

Comment:
```suggestion
   patab$addFootnote(message = footnote)
```

Reply:
I moved that up, because I was too lazy to suggest changes line by line :) Sorry

Reply:
No problem at all! It's definitely cleaner.